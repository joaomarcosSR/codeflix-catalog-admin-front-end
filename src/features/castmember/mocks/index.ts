export const castMemberResponse = {
  current_page: 0,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      id: "27840c5ce79b40bd8f160f5b0e31e776",
      name: "Amanda",
      type: "ACTOR",
      created_at: "2023-01-05T10:34:23.154624Z",
    },
    {
      id: "f17e54f18bad43e0a1bfa085dcfab954",
      name: "Andreia",
      type: "ACTOR",
      created_at: "2023-01-05T10:35:45.222616Z",
    },
    {
      id: "789f5ed13e8a47c5a897386f6884b562",
      name: "Gabriel",
      type: "ACTOR",
      created_at: "2023-01-05T10:35:26.799065Z",
    },
    {
      id: "464678cff4da4125bf8946817fb51105",
      name: "Jordana",
      type: "ACTOR",
      created_at: "2023-01-05T10:35:54.720426Z",
    },
    {
      id: "e134c302d7dc413ca7319ed1306a1b36",
      name: "Julia",
      type: "ACTOR",
      created_at: "2022-12-24T00:51:55.768063Z",
    },
    {
      id: "aae07ae17e3a44b59837ba220b8e2f0b",
      name: "Leticia",
      type: "ACTOR",
      created_at: "2023-01-05T10:34:30.448295Z",
    },
    {
      id: "5e9efad9fc444630bb7a53464449f422",
      name: "Marcos",
      type: "ACTOR",
      created_at: "2023-01-05T10:35:00.398915Z",
    },
    {
      id: "349397fef45d46b08f0e35277229754a",
      name: "Maria",
      type: "DIRECTOR",
      created_at: "2023-01-05T10:34:38.171395Z",
    },
    {
      id: "d80d748e70914491af69e6192dddcb22",
      name: "Paulo",
      type: "ACTOR",
      created_at: "2023-01-05T10:36:04.139962Z",
    },
    {
      id: "c251ae2593a74b4e851d760aa532a24d",
      name: "Pedro",
      type: "DIRECTOR",
      created_at: "2022-12-24T00:49:09.326592Z",
    },
  ],
};

export const castMemberResponsePage2 = {
  current_page: 1,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      id: "50271aacf218449ea01cca393ff45a24",
      name: "Teresinha",
      type: "DIRECTOR",
      created_at: "2023-01-05T10:35:36.934111Z",
    },
  ],
};
