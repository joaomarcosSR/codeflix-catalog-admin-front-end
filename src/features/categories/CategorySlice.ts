import {
  Category,
  CategoryParams,
  CreateCategoryResult,
  ListCategoryResult,
  UpdateCategoryResult,
} from "../../types/Category";
import { apiSlice } from "../api/apiSlice";

const endpointUrl = "/categories";

export const initialState: Category = {
  id: "",
  name: "",
  description: "",
  is_active: false,
  created_at: "",
  deleted_at: null,
};

function parseQueryParams(params: CategoryParams): string {
  const query = new URLSearchParams();

  if (params.page) {
    query.append("page", params.page.toString());
  }

  if (params.perPage) {
    query.append("per_page", params.perPage.toString());
  }

  if (params.search) {
    query.append("search", params.search);
  }

  return query.toString();
}

function createCategoryMutation(category: Category) {
  return { url: endpointUrl, method: "POST", body: category };
}

function getCategoryById(id: string): string {
  return `${endpointUrl}/${id}`;
}

function getCategories({ page = 0, perPage = 10, search = "" }): string {
  const params = { page, perPage, search };
  // console.log(`${endpointUrl}?${parseQueryParams(params)}`);
  return `${endpointUrl}?${parseQueryParams(params)}`;
}

function updateCategoryMutation(category: Category) {
  return {
    url: `${endpointUrl}/${category.id}`,
    method: "PUT",
    body: category,
  };
}

function deleteCategoryMutation(category: Category) {
  return {
    url: `${endpointUrl}/${category.id}`,
    method: "DELETE",
  };
}

export const categoriesApiSlice = apiSlice.injectEndpoints({
  endpoints: ({ query, mutation }) => ({
    createCategory: mutation<CreateCategoryResult, Category>({
      query: createCategoryMutation,
      invalidatesTags: ["Categories"],
    }),
    getCategoryById: query<Category, string>({
      query: getCategoryById,
      providesTags: ["Categories"],
    }),
    getCategories: query<ListCategoryResult, CategoryParams>({
      query: getCategories,
      providesTags: ["Categories"],
    }),
    updateCategory: mutation<UpdateCategoryResult, Category>({
      query: updateCategoryMutation,
      invalidatesTags: ["Categories"],
    }),
    deleteCategory: mutation<void, { id: string }>({
      query: deleteCategoryMutation,
      invalidatesTags: ["Categories"],
    }),
  }),
});

export const {
  useCreateCategoryMutation,
  useGetCategoriesQuery,
  useGetCategoryByIdQuery,
  useUpdateCategoryMutation,
  useDeleteCategoryMutation,
} = categoriesApiSlice;
