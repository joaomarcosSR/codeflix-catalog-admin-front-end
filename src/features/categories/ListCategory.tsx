import { Box, Button, Typography } from "@mui/material";
import { GridFilterModel } from "@mui/x-data-grid";
import { useSnackbar } from "notistack";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import {
  useDeleteCategoryMutation,
  useGetCategoriesQuery,
} from "./CategorySlice";
import { CategoriesTable } from "./components/CategoryTable";

export const CategoryList = () => {
  const [options, setOptions] = useState({
    page: 0,
    perPage: 10,
    search: "",
    rowsPerPage: [10, 25, 50, 100],
  });

  const { data, isFetching, error } = useGetCategoriesQuery(options);
  const [deleteCategory, deleteCategoryStatus] = useDeleteCategoryMutation();
  const { enqueueSnackbar } = useSnackbar();

  function handleOnPageChange(page: number) {
    setOptions({ ...options, page });
  }

  function handleOnPageSizeChange(perPage: number) {
    setOptions({ ...options, perPage });
  }

  function handleFilterChange(filterModel: GridFilterModel) {
    const search = filterModel.quickFilterValues?.length
      ? filterModel.quickFilterValues.join(" ")
      : "";
    setOptions({ ...options, search });
  }

  async function handleDeleteCategory(id: string) {
    await deleteCategory({ id });
  }

  useEffect(() => {
    if (deleteCategoryStatus.isSuccess) {
      enqueueSnackbar("Category deleted successfully", { variant: "success" });
    } else if (deleteCategoryStatus.error) {
      enqueueSnackbar("Category not deleted", { variant: "error" });
    }
  }, [deleteCategoryStatus, enqueueSnackbar]);

  if (error) {
    return <Typography>Error fetching categories</Typography>;
  }

  return (
    <Box maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
      <Box display="flex" justifyContent="flex-end">
        <Button
          variant="contained"
          color="secondary"
          component={Link}
          to="/categories/create"
          style={{ marginBottom: "1rem" }}
        >
          + New
        </Button>
      </Box>

      <CategoriesTable
        data={data}
        perPage={options.perPage}
        rowsPerPage={options.rowsPerPage}
        isFetching={isFetching}
        handleOnPageChange={handleOnPageChange}
        handleFilterChange={handleFilterChange}
        handleOnPageSizeChange={handleOnPageSizeChange}
        handleDelete={handleDeleteCategory}
      />
    </Box>
  );
};
