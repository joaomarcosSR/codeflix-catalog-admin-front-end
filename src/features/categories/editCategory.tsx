import { Box, Paper, Typography } from "@mui/material";
import { useSnackbar } from "notistack";
import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { Category } from "../../types/Category";
import {
  defaultCategory,
  useGetCategoryByIdQuery,
  useUpdateCategoryMutation,
} from "./CategorySlice";
import { CategoryForm } from "./components/CategoryForm";

export const CategoryEdit = () => {
  const id = useParams().id || "";
  const { data: category, isFetching } = useGetCategoryByIdQuery(id);
  const [updateCategory, updateCategoryStatus] = useUpdateCategoryMutation();
  const [categoryState, setCategoryState] = useState<Category>(
    defaultCategory()
  );
  const { enqueueSnackbar } = useSnackbar();

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    await updateCategory(categoryState);
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    console.log(name, value);
    setCategoryState({ ...categoryState, [name]: value });
  };
  const handleToggle = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, checked } = event.target;
    setCategoryState({ ...categoryState, [name]: checked });
  };

  useEffect(() => {
    if (category) {
      setCategoryState(category);
    }
  }, [category]);

  useEffect(() => {
    if (updateCategoryStatus.isSuccess) {
      enqueueSnackbar("Category updated successfully", { variant: "success" });
    }
    if (updateCategoryStatus.error) {
      enqueueSnackbar(
        `Category not updated. Error: ${updateCategoryStatus.error}`,
        { variant: "error" }
      );
    }
  }, [updateCategoryStatus, enqueueSnackbar]);

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Edit Category</Typography>
          </Box>
        </Box>

        <CategoryForm
          category={categoryState}
          isLoading={isFetching}
          isDisabled={updateCategoryStatus.isLoading}
          handleSubmit={handleSubmit}
          handleChange={handleChange}
          handleToggle={handleToggle}
        />
      </Paper>
    </Box>
  );
};
