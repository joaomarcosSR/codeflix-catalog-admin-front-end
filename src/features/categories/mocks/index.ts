export const categoryResponse = {
  current_page: 0,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      is_active: true,
      created_at: "2022-12-23T13:53:22.999118Z",
      deleted_at: null,
      id: "1e3ac7c57842470383f70daf8ee22bc8",
      name: "Antonio",
      description: "Ficar de olho",
    },
    {
      is_active: true,
      created_at: "2022-12-22T21:03:15.015422Z",
      deleted_at: null,
      id: "3432d944c25948738e406b9589532e73",
      name: "Best Series",
      description: "The best series",
    },
    {
      is_active: true,
      created_at: "2022-12-23T13:52:54.363829Z",
      deleted_at: null,
      id: "82bd2788d209498cb39f4f4181d85c50",
      name: "Joao",
      description: "Meu nome",
    },
    {
      is_active: true,
      created_at: "2022-12-23T14:15:06.198204Z",
      deleted_at: null,
      id: "c2ff7178bd36467587c0097fb7d101d6",
      name: "Olives",
      description: "Do curso",
    },
    {
      is_active: true,
      created_at: "2022-12-23T13:53:22.999118Z",
      deleted_at: null,
      id: "1e3ac7c57842470383f70daf8ee85bc8",
      name: "Carlos",
      description: "Ficar de olho",
    },
    {
      is_active: true,
      created_at: "2022-12-22T21:03:15.015422Z",
      deleted_at: null,
      id: "3432d944c25945638e406b9589532e73",
      name: "Worst Series",
      description: "The worst series",
    },
    {
      is_active: true,
      created_at: "2022-12-23T13:52:54.363829Z",
      deleted_at: null,
      id: "82bd2788d209478cb39f4f4181d85c50",
      name: "Marcos",
      description: "Meu nome",
    },
    {
      is_active: true,
      created_at: "2022-12-23T14:15:06.198204Z",
      deleted_at: null,
      id: "c2ff7178tr36467587c0097fb7d101d6",
      name: "Wesley",
      description: "Do curso",
    },
    {
      is_active: true,
      created_at: "2022-12-23T13:53:22.999118Z",
      deleted_at: null,
      id: "1e3ac7c57842250383f70daf8ee22bc8",
      name: "Silva",
      description: "Ficar de olho",
    },
    {
      is_active: true,
      created_at: "2022-12-22T21:03:15.015422Z",
      deleted_at: null,
      id: "3432d944c25948765e406b9589532e73",
      name: "Best Films",
      description: "The best films",
    },
  ],
};

export const categoryResponsePage2 = {
  current_page: 1,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      is_active: true,
      created_at: "2022-12-23T14:15:06.198204Z",
      deleted_at: null,
      id: "c2ff7178bd36467587c0097fb7d101d6",
      name: "Armandro",
      description: "Do curso",
    },
  ],
};
