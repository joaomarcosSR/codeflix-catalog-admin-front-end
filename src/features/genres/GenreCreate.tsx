import { Paper, Typography } from "@mui/material";
import { Box } from "@mui/system";
import { useSnackbar } from "notistack";
import { useEffect, useState } from "react";
import { Genre } from "../../types/Genre";
import { GenreForm } from "./components/GenreForm";
import {
  initialState,
  useCreateGenreMutation,
  useGetGenreCategoriesQuery,
} from "./GenreSlice";

export const GenreCreate = () => {
  const [isDisabled, setIsDisabled] = useState(false);
  const { enqueueSnackbar } = useSnackbar();
  const [search, setSearch] = useState("");
  const { data: categoriesResult } = useGetGenreCategoriesQuery(search);
  const [createGenre, status] = useCreateGenreMutation();
  const [genreState, setGenreState] = useState<Genre>(initialState);

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    await createGenre(genreState);
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setGenreState({ ...genreState, [name]: value });
  };

  function handleFilterChange(typed: string): void {
    setSearch(typed || "");
  }

  const handleToggle = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, checked } = event.target;
    setGenreState({ ...genreState, [name]: checked });
  };

  useEffect(() => {
    if (status.isSuccess) {
      enqueueSnackbar("Genre created successfully", {
        variant: "success",
      });
      setIsDisabled(true);
    }
    if (status.error) {
      enqueueSnackbar("Genre not created", { variant: "error" });
      setIsDisabled(false);
    }
  }, [status, enqueueSnackbar]);

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Create Genre</Typography>
          </Box>
        </Box>
        <GenreForm
          genre={genreState}
          categories={categoriesResult?.items}
          isLoading={status.isLoading}
          isDisabled={isDisabled}
          handleSubmit={handleSubmit}
          handleChange={handleChange}
          handleAutoCompleteFilterChange={handleFilterChange}
          handleToggle={handleToggle}
        />
      </Paper>
    </Box>
  );
};
