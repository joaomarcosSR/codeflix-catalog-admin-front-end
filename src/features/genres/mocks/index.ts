export const genreResponse = {
  current_page: 0,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      is_active: true,
      created_at: "2023-01-07T21:50:10.467401Z",
      deleted_at: null,
      id: "ff4a68dc7b4f4d05be4ae1c75f67f75e",
      name: "Ação",
    },
    {
      is_active: true,
      created_at: "2023-01-07T11:58:30.300105Z",
      deleted_at: null,
      id: "b4d94a9aa55f4e60aab02bb86afbeff0",
      name: "Drama",
    },
    {
      is_active: true,
      created_at: "2023-01-07T22:21:45.530717Z",
      deleted_at: null,
      id: "7bb64ebcabdb413ca84f2264eaf1eff5",
      name: "Genero 1",
    },
    {
      is_active: true,
      created_at: "2023-01-07T22:21:58.428416Z",
      deleted_at: null,
      id: "b11ca00ce01c4adbad1fd52b343128f1",
      name: "Genero 2",
    },
    {
      is_active: true,
      created_at: "2023-01-07T22:22:10.970088Z",
      deleted_at: null,
      id: "d143f9e7fc4f449ba19c6adbe12a2e72",
      name: "Genero 3",
    },
    {
      is_active: false,
      created_at: "2023-01-07T22:22:21.795369Z",
      deleted_at: "2023-01-07T22:22:21.795369Z",
      id: "77c89ee45d274883aacddfbf46f7a877",
      name: "Genero 4",
    },
    {
      is_active: true,
      created_at: "2023-01-07T22:22:48.868797Z",
      deleted_at: null,
      id: "7c3142e2360f429abd06fb0908b5cd36",
      name: "Genero 5",
    },
    {
      is_active: false,
      created_at: "2023-01-07T22:23:00.961794Z",
      deleted_at: "2023-01-07T22:23:00.961794Z",
      id: "0789821cdd43427d98f0377a4c267228",
      name: "Genero 6",
    },
    {
      is_active: true,
      created_at: "2023-01-07T11:57:01.677036Z",
      deleted_at: null,
      id: "f75e77648ac5480db86baf40282a4ff4",
      name: "Humor",
    },
    {
      is_active: false,
      created_at: "2023-01-07T01:44:57.368521Z",
      deleted_at: "2023-01-07T01:44:57.368521Z",
      id: "ce356ef0762248e6aea7cafb311dc495",
      name: "Terror",
    },
  ],
};

export const genreResponsePage2 = {
  current_page: 1,
  per_page: 10,
  total_pages: 2,
  total_elements: 11,
  items: [
    {
      is_active: true,
      created_at: "2023-01-07T22:21:19.460325Z",
      deleted_at: null,
      id: "90ad18b8a38944a29593e2fec3af379c",
      name: "Test",
    },
  ],
};
