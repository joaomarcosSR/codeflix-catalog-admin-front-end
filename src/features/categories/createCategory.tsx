import { Box, Paper, Typography } from "@mui/material";
import { useSnackbar } from "notistack";
import { useState } from "react";
import { useAppDispatch } from "../../app/hooks";
import { Category, createCategory } from "./CategorySlice";
import { CategoryForm } from "./components/CategoryForm";

export const CategoryCreate = () => {
  const [isDisabled, setIsDisabled] = useState(false);
  const [categoryState, setCategoryState] = useState<Category>({
    id: "",
    name: "",
    description: "",
    is_active: false,
    created_at: "",
    deleted_at: "",
  });
  const dispatch = useAppDispatch();
  const { enqueueSnackbar } = useSnackbar();

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    dispatch(createCategory(categoryState));
    enqueueSnackbar("Category created successfully", { variant: "success" });
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setCategoryState({ ...categoryState, [name]: value });
  };
  const handleToggle = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, checked } = event.target;
    setCategoryState({ ...categoryState, [name]: checked });
  };

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Create Category</Typography>
          </Box>
        </Box>

        <CategoryForm
          category={categoryState}
          isDisabled={isDisabled}
          handleSubmit={handleSubmit}
          handleChange={handleChange}
          handleToggle={handleToggle}
        />
      </Paper>
    </Box>
  );
};
