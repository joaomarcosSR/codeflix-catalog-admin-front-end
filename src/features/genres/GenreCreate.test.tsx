import { rest } from "msw";
import { setupServer } from "msw/node";

import {
  fireEvent,
  renderWithProviders,
  screen,
  waitFor,
} from "../../utils/test-utils";
import { baseUrl } from "../api/apiSlice";
import { categoryResponse } from "../categories/mocks";
import { GenreCreate } from "./GenreCreate";

export const handlers = [
  rest.get(`${baseUrl}/categories`, (_, res, ctx) => {
    return res(ctx.json(categoryResponse));
  }),
  rest.post(`${baseUrl}/genres`, (_, res, ctx) => {
    return res(ctx.delay(150), ctx.status(201));
  }),
];

const server = setupServer(...handlers);

describe("Create Genre", () => {
  afterAll(() => server.close());
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());

  it("should render correctly", () => {
    const { asFragment } = renderWithProviders(<GenreCreate />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle submit success", async () => {
    renderWithProviders(<GenreCreate />);

    const name = screen.getByTestId("name");
    const submit = screen.getByRole("button", { name: /Save/i });

    await waitFor(() => {
      expect(submit).toBeInTheDocument();
    });

    fireEvent.change(name, { target: { value: "Test" } });
    fireEvent.click(submit);

    await waitFor(() => {
      const text = screen.getByText("Genre created successfully");
      expect(text).toBeInTheDocument();
    });
  });

  it("should handle submit error", async () => {
    server.use(
      rest.post(`${baseUrl}/genres`, (_, res, ctx) => {
        return res(ctx.status(500));
      })
    );

    renderWithProviders(<GenreCreate />);

    const name = screen.getByTestId("name");
    const submit = screen.getByRole("button", { name: /Save/i });

    await waitFor(() => {
      expect(submit).toBeInTheDocument();
    });

    fireEvent.change(name, { target: { value: "Test" } });
    fireEvent.click(submit);

    await waitFor(() => {
      const text = screen.getByText("Genre not created");
      expect(text).toBeInTheDocument();
    });
  });
});
