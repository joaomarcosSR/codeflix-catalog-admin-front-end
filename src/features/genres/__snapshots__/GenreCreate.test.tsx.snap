// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Genre should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-19midj6"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Create Genre
          </h4>
        </div>
      </div>
      <div
        class="MuiBox-root css-19midj6"
      >
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r0:"
                    id=":r0:-label"
                  >
                    Name
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                       *
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                      data-testid="name"
                      id=":r0:"
                      name="name"
                      required=""
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-yjsfm1"
                      >
                        <span>
                          Name *
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  data-testid="categories-input"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r1:"
                    id=":r1:-label"
                  >
                    Categories
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                      disabled=""
                      id=":r1:"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        disabled=""
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-yjsfm1"
                      >
                        <span>
                          Categories
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
              >
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
                  >
                    <span
                      class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary css-priqlz-MuiButtonBase-root-MuiSwitch-switchBase"
                      data-testid="is_active"
                    >
                      <input
                        aria-label="controlled"
                        class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                        name="is_active"
                        type="checkbox"
                      />
                      <span
                        class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
                      />
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                    <span
                      class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
                    />
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                  >
                    Active
                  </span>
                </label>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-10egq61"
              >
                <a
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                  href="/genres"
                  tabindex="0"
                >
                  Back
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </a>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="submit"
                >
                  Save
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
